kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -18,94 +18,276 @@
+ *
+ * This class exposes only two public APIs
+ * - {@link #lazyLoadProtocolAndMetadata}: replay the log in reverse and return the first non-null
+ *                                         Protocol and Metadata
+ * - {@link #getAddFilesAsColumnarBatches}: return all active (not tombstoned) AddFiles as
+ *                                          {@link ColumnarBatch}s
+ */
+public class LogReplay implements Logging {
+
+    /** Read schema when searching for the latest Protocol and Metadata. */
+    public static final StructType PROTOCOL_METADATA_READ_SCHEMA = new StructType()
+        .add("protocol", Protocol.READ_SCHEMA)
+        .add("metaData", Metadata.READ_SCHEMA);
+
+    /** Read schema when searching for all the active AddFiles (need the RemoveFiles, too). */
+    public static final StructType ADD_REMOVE_READ_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA)
+        .add("remove", RemoveFile.READ_SCHEMA);
+
+    /** Data (result) schema of the remaining active AddFiles. */
+    public static final StructType ADD_ONLY_DATA_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA);
+
     private final Path dataPath;
     private final LogSegment logSegment;
-    private final CloseableIterable<Tuple2<Action, Boolean>> reverseActionsIterable;
+    private final TableClient tableClient;
+
+    /**
+     * An iterable generating iterators of ColumnarBatches of AddFiles and RemovedFiles, used to
+     * return the active AddFiles in the Snapshot.
+     */
+    private final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
+        addRemoveColumnarBatchIterable;
+
     private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
 
     public LogReplay(
-        Path logPath,
-        Path dataPath,
-        TableClient tableHelper,
-        LogSegment logSegment)
-    {
+            Path logPath,
+            Path dataPath,
+            TableClient tableClient,
+            LogSegment logSegment) {
         this.dataPath = dataPath;
         this.logSegment = logSegment;
+        this.tableClient = tableClient;
 
-        final Stream<FileStatus> allFiles = Stream.concat(
-            logSegment.checkpoints.stream(),
-            logSegment.deltas.stream());
+        final List<FileStatus> allFiles = Stream.concat(
+                logSegment.checkpoints.stream(),
+                logSegment.deltas.stream()
+            ).collect(Collectors.toList());
         assertLogFilesBelongToTable(logPath, allFiles);
 
-        this.reverseActionsIterable = new ReverseFilesToActionsIterable(
-            tableHelper,
-            allFiles);
-        this.protocolAndMetadata = new Lazy<>(this::loadTableProtocolAndMetadata);
+        this.addRemoveColumnarBatchIterable = new ReverseFilesToColumnarBatchActionsIterable(
+            tableClient,
+            allFiles,
+            ADD_REMOVE_READ_SCHEMA);
+
+        final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
```

Comment:
super nit: can this be put inside the () -> { ... }  below since that is the only thing using this object? wouldnt that be less code?